### PR TITLE
Flags for SIOC{G,S}IFFLAGS

### PR DESCRIPTION
**Pull request by bentheklutz**
Adds a packed struct for SIOCGIFFLAGS and SIOCSIFFLAGS

Addresses #21890
Quick checked this with the following code.

```
const std = @import("std");

pub fn main() !void {
    const sock = try std.posix.socket(std.posix.AF.INET, std.posix.SOCK.STREAM, 0);
    const ifname = "wlp0s20f3";
    var ifr: std.posix.ifreq = std.mem.zeroInit(std.posix.ifreq, .{});
    @memcpy(ifr.ifrn.name[0..ifname.len], ifname);
    ifr.ifrn.name[ifname.len] = 0;
    try std.posix.ioctl_SIOCGIFINDEX(sock, &ifr);
    const rval = std.posix.errno(std.os.linux.ioctl(sock, std.os.linux.SIOCGIFFLAGS, @intFromPtr(&ifr)));
    if (rval != .SUCCESS) {
        return error.FlagsGetFailed;
    }

    std.debug.print("Flags set:\n", .{});

    if (ifr.ifru.flags.UP) {
        std.debug.print("UP\n", .{});
    }
    if (ifr.ifru.flags.BROADCAST) {
        std.debug.print("BROADCAST\n", .{});
    }
    if (ifr.ifru.flags.DEBUG) {
        std.debug.print("DEBUG\n", .{});
    }
    if (ifr.ifru.flags.LOOPBACK) {
        std.debug.print("LOOPBACK\n", .{});
    }
    if (ifr.ifru.flags.POINTOPOINT) {
        std.debug.print("POINTOPOINT\n", .{});
    }
    if (ifr.ifru.flags.NOTRAILERS) {
        std.debug.print("NOTRAILERS\n", .{});
    }
    if (ifr.ifru.flags.RUNNING) {
        std.debug.print("RUNNING\n", .{});
    }
    if (ifr.ifru.flags.NOARP) {
        std.debug.print("NOARP\n", .{});
    }
    if (ifr.ifru.flags.PROMISC) {
        std.debug.print("PROMISC\n", .{});
    }

    // requires elevated permissions to execute
    // ifr.ifru.flags.PROMISC = true;
    // ifr.ifru.flags.BROADCAST = true;

    // rval = std.posix.errno(std.os.linux.ioctl(sock, std.os.linux.SIOCSIFFLAGS, @intFromPtr(&ifr)));
    // switch (rval) {
    //     .SUCCESS => {},
    //     .PERM => return error.FailedPermissions,
    //     else => {},
    // }
}

```

### CODE REVIEW

File: lib/std/os/linux.zig [hunk 1]
```diff
@@ -3987,6 +3987,18 @@ pub const IP = struct {
...
+    UP: bool = false,
+    BROADCAST: bool = false,
+    DEBUG: bool = false,
+    LOOPBACK: bool = false,
+    POINTOPOINT: bool = false,
+    NOTRAILERS: bool = false,
+    RUNNING: bool = false,
+    NOARP: bool = false,
+    PROMISC: bool = false,
+};
```
Comment (bentheklutz) [inline lib/std/os/linux.zig:4000]: Forgot to fill out the rest of the bits.
```suggestion
    _: u7 = 0,
};
```

**Comment by bentheklutz** [comment]
Should be actually good now. I didn't configure zig to use std from the compiler's source tree, so I was making changes in two places and not keeping them in sync.

File: lib/std/os/linux.zig [hunk 2]
```diff
@@ -7121,6 +7121,19 @@ pub const SIOCPROTOPRIVATE = 0x89E0;
...
+    UP: bool = false,
+    BROADCAST: bool = false,
+    DEBUG: bool = false,
+    LOOPBACK: bool = false,
+    POINTOPOINT: bool = false,
+    NOTRAILERS: bool = false,
+    RUNNING: bool = false,
+    NOARP: bool = false,
+    PROMISC: bool = false,
+    _: u7 = 0,
```
Comment (alexrp) [inline lib/std/os/linux.zig:7134]:
```suggestion
    _9: u7 = 0,
```

**Review by alexrp — APPROVED** [review]

File: lib/std/os/linux.zig (hunk 2 above)
Comment (gryzus24) [reply lib/std/os/linux.zig]: Hello, is there any particular reason that the remaining flags shouldn't be exported? `man netdevice` lists a lot more of them and they are set just fine. Here is a `std.debug.print` output on IFF on my interface:
```
os.linux.IFF{ .UP = true, .BROADCAST = true, .DEBUG = false, .LOOPBACK = false, .POINTOPOINT = false, .NOTRAILERS = false, .RUNNING = true, .NOARP = false, .PROMISC = false, ._9 = 8 }
```
notice the `._9 = 8` which indicates that IFF_MULTICAST is set and I can toggle it just fine with `sudo ip link set eth0 multicast on/off`.

I suspect why you didn't export the remaining flags; the glibc include/net/if.h header lists them as "Not supported." (musl does not comment on them - just exports them as the kernel does. Also, notice the order of these flags in the `man netdevice` manual differs from what the kernel exports in its enum).

The "Not supported." comment in glibc source code goes as far as the github mirror archives go (https://github.com/bminor/glibc/commit/df4ef2ab9c0899b2670067cd97e58f7eb2913e00#diff-81b12b4ec3b272eaf8c4283f435570546dd194f8d62c431180ed2d0c0cf683acR39) and I sense the reason this comment is in place is a lot more mundane than it might seem, but I won't hazard a guess :-)

Also, the driver for my network interface has code paths to handle the "Not supported" IFF_ALLMULTI flag.

I speak with no authority on the subject matter. The thing is that this change broke my code, so please interpret my inquiry as nothing but mild indignation with a tinge of curiosity :-)

Comment (alexrp) [reply lib/std/os/linux.zig]: If the other flags are in fact useful then I don't see any reason why we wouldn't want to add them.

Comment (gryzus24) [reply lib/std/os/linux.zig]: For anyone wondering what's the history of this comment.

- glibc up to 1996 was using the Linux kernel generated header for net/if.h, after that they just copied the kernel header that contained this comment into the glibc source tree.
- before 1996, in the Linux kernel release 1.1.69, the comment was changed from "These are not yet used:" to "Not supported" - these were just predefined macros that were not used at the time. See: https://repo.or.cz/davej-history.git/blobdiff/fb3ed0dc9690ef62209db23d60ab415511d6734f..ea8a68b948397fa9cd1c8a1a6b61a4dc4bc99ec5:/include/linux/if.h

So basically Zig exports the state of IF flags as defined in Linux in 1994. The glibc header is just stale.
